android/src/main/java/com/getbouncer/RNCardVerifyModule.java

```diff
@@ -306,11 +311,44 @@ public void isSupportedAsync(Promise promise) {
     }
 
     @ReactMethod
-    public void downloadModels() {
+    public void awaitReady(@NotNull final Promise promise) {
+        if (modelsDownloaded) {
+            promise.resolve(true);
+        } else {
+            modelsDownloadPromise = promise;
```

Comment:
Depending on how you handle threads with `prepareScan` there is the possibility of a race condition where this might get dropped.

Reply:
I actually thought about this while writing this code, but I'm not sure react-native gives us a good solution. The [react-native docs on threading](https://reactnative.dev/docs/native-modules-android#threading) indicate that all `@ReactMethod` functions execute on the same thread, but we have no guarantee what thread that actually is. The same docs also [strongly discourage synchronized methods](https://reactnative.dev/docs/native-modules-android#synchronous-methods) because they require substantial memory usage changes.

The documentation on the methods used to invoke the completion callback, and ["There is no guarantee on the execution context in which the handler is invoked."](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/invoke-on-completion.html)

Given these limits, I'm not sure there's much that we can do to guarantee no race condition here. At the very least, I've ordered the conditionals in such a way as to minimize the possibility of one.

Reply:
Looks like after some more research, I may be able to safely pin both of these to the UI thread, which should resolve any race conditions. I'm still a little unclear on the implications of doing so, but in my testing, it's working.

Reply:
Now I'm worried about deadlock! Honestly +1000 to there not being a great solution here, whatever path you pick sounds good to me.